river/covariance/emp.py

```diff
@@ -178,35 +194,63 @@ def update_many(self, X: pd.DataFrame):
         mean_arr = X_arr.mean(axis=0)
         cov_arr = np.cov(X_arr.T, ddof=self.ddof)
 
+        n = len(X)
         mean = dict(zip(X.columns, mean_arr))
         cov = {
             (i, j): cov_arr[r, c]
             for (r, i), (c, j) in itertools.combinations_with_replacement(enumerate(X.columns), r=2)
         }
 
-        for i, j in itertools.combinations(sorted(X.columns), r=2):
+        self = self._from_state(n=n, mean=mean, cov=cov, ddof=self.ddof)
+
+        return self
+
+    @classmethod
+    def _from_state(cls, n: int, mean: dict, cov: dict, *, ddof=1):
+        """Create a new instance from state information.
+
+        Parameters
+        ----------
+        cls (type)
+            The class type.
+        n (int)
+            The number of data points.
+        mean (dict)
+            A dictionary of variable means.
+        cov (dict)
+            A dictionary of covariance or variance values.
+        ddof (int, optional)
+            Degrees of freedom for covariance calculation. Defaults to 1.
+
+        Returns
+        ----------
+            cls: A new instance of the class with updated covariance matrix.
+
+        Raises
+        ----------
+            KeyError: If an element in `mean` or `cov` is missing.
+        """
+        new = cls(ddof=ddof)
+        for i, j in itertools.combinations(sorted(mean.keys()), r=2):
             try:
-                self[i, j]
+                new[i, j]
             except KeyError:
-                self._cov[i, j] = stats.Cov(self.ddof)
-            self._cov[i, j] += stats.Cov._from_state(
-                n=len(X),
+                new._cov[i, j] = stats.Cov(new.ddof)
+            new._cov[i, j] += stats.Cov._from_state(
+                n=n,
                 mean_x=mean[i],
                 mean_y=mean[j],
                 cov=cov.get((i, j), cov.get((j, i))),
-                ddof=self.ddof,
+                ddof=new.ddof,
             )
 
-        for i in X.columns:
+        for i in sorted(mean.keys()):
```

Comment:
Is this sorting operation a must? I think we can save a couple of milliseconds and bring no impacts, functionality-wise. What do you think @MarekWadinger?

Reply:
The same applies to the previous for loop (the one taking combinations of the x's keys.